### ISSUE TITLE
Use box drawing characters?

### ISSUE THREAD

**Issue by Emerentius** [XAMPPRocky/tokei#249]
Skimming through the new comparison between tools, I noticed how all of them use the visually noisy hyphen `-----` lines. The [box drawing characters](https://en.wikipedia.org/wiki/Box-drawing_character) could be used to have clean, uninterrupted lines. These improve aesthetics and improve readability.
Of course, this requires unicode support. I wasn't sure if the choice of hyphens was deliberate to support legacy terminals.

Some comparisons:
```
-----------------------------------------------------------------------------------
 Language                Files        Lines         Code     Comments       Blanks
-----------------------------------------------------------------------------------
 Dockerfile                 14          251          219            6           26
 Markdown                    3          180          180            0            0
 Rust                      102        41073        28770         8630         3673
 Shell                       6          310          195           63           52
 TOML                        8          141          120            3           18
 XML                         1       134861       133755            0         1106
-----------------------------------------------------------------------------------
 Total                     134       176816       163239         8702         4875
-----------------------------------------------------------------------------------
```

```
───────────────────────────────────────────────────────────────────────────────────
 Language                Files        Lines         Code     Comments       Blanks
───────────────────────────────────────────────────────────────────────────────────
 Dockerfile                 14          251          219            6           26
 Markdown                    3          180          180            0            0
 Rust                      102        41073        28770         8630         3673
 Shell                       6          310          195           63           52
 TOML                        8          141          120            3           18
 XML                         1       134861       133755            0         1106
───────────────────────────────────────────────────────────────────────────────────
 Total                     134       176816       163239         8702         4875
───────────────────────────────────────────────────────────────────────────────────
```

Following LaTeX `booktabs` style:
```
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  Language                Files        Lines         Code     Comments       Blanks
───────────────────────────────────────────────────────────────────────────────────
 Dockerfile                 14          251          219            6           26
 Markdown                    3          180          180            0            0
 Rust                      102        41073        28770         8630         3673
 Shell                       6          310          195           63           52
 TOML                        8          141          120            3           18
 XML                         1       134861       133755            0         1106
───────────────────────────────────────────────────────────────────────────────────
 Total                     134       176816       163239         8702         4875
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

**Comment by Aaronepower** [comment]
Thank you for this issue! It is definitely deliberate on my part, I would however accept a PR that checked if the terminal supported unicode and used the box characters then and reverted back to the hyphens if it wasn't present.

**Comment by Aaronepower** [comment]
The relevant code currently is just this single line. https://github.com/Aaronepower/tokei/blob/cd26407645562dc12b7e468b42c7a138d64b61ab/src/main.rs#L198

**Comment by Emerentius** [comment]
Afaik it is not actually possible to test whether a certain character will be correctly printed (e.g. font may miss it).
I could do a best efforts approach. Assume unicode capability, if the codeset in `$LANG` is set to UTF-8. That should cover all correctly configured POSIX-y platforms. `tree` does it this way.
For windows I could maybe test it by printing a multi-byte, one wide character and check the cursor location.

**Comment by Aaronepower** [comment]
@Emerentius Hmm, maybe that should be a seperate crate so this could be reused across crates.

**Comment by Aaronepower** [comment]
@Emerentius That code looks fine overall. Once you've done the testing that this works for the big three platforms and published on crates.io, feel free to make a PR!

**Comment by Aaronepower** [comment]
@Emerentius Hello, how has testing the prototype gone, have you tested it on all major platforms?

**Comment by Emerentius** [comment]
Oh, I forgot about this one, sorry

I had done some tests on a few windows systems. I have PCs here with windows 7, 8.1 and 10. The thin box drawing characters worked everywhere, the thick ones did not, but I don't think we can rely on that. Windows uses various charsets (codepages as they call them) depending on locale and some may not have box drawing characters. In German versions, cmd.exe uses codepage 850 which does have them. There seems to be a translation layer somewhere in the printing pipeline. A UTF-8 line drawing character is 3 bytes but only 1 byte in codepage 850.
The test for unicode support by trial printing I've mentioned above is not reliable, because the windows console will just replace unknown characters with a question mark or a box resulting in false positives on the width check.

I also tested git bash on windows which kinda worked. `$LANG` was always set to `de_DE.UTF-8`. Thick lines again made problems, resulting in `?` on windows 7 and just regular ascii `-` on windows 10. Thin lines worked.

I have no issues under Ubuntu. I don't have a mac to test on.

The TL;DR is, testing support doesn't seem to be possible, but relying on the explicitly set charset seems reasonable. If the terminal claims to support unicode by way of `$LANG` being set to `UTF-8`, we could use box drawing characters. If that fails, it's a font issue.
The thick lines aren't as well supported and imo the LaTeX style doesn't look that good with bright lines on dark background, anyway. I'd say we should just go with only thin lines.

**Comment by throwaway1037** [comment]
Perhaps there could be CLI options implemented to switch between the current hyphen tables, box drawing characters, markdown tables, or LaTeX tables, eg. `-T` or `--table-style` with arguments such as `comp`, `box`, `md`, or `tex`.

**Comment by XAMPPRocky** [comment]
Well for more generic table support, I think I'd prefer to add CSV as an output option, and then you could pipe the CSV output into any of the CSV transform tools that can output it into another format.

**Comment by throwaway1037** [comment]
At the moment the `-o` option supports output in cbor, json, and yaml. Would a generic implementation which could output in all `serde` supported formats be possible?

I don't know whether it would be best to just support all output formats directly or choose to only support the one which includes the most information, and then pipe the data to a conversion tool.

**Comment by XAMPPRocky** [comment]
> Would a generic implementation which could output in all serde supported formats be possible?

Tokei wouldn't need to support all the formats, only CSV. Then you can use pipes in your shell to convert into something else, for example if CSV was supported you could write this to convert from the CSV into LaTeX tables using [tably](https://github.com/narimiran/tably).

```zsh
python tably.py <(tokei ~/src -o csv)
```

**Comment by throwaway1037** [comment]
If tokei is going down that route, then shouldn't it be best to select the format which encodes the most information, so that the output is not lossy?

I think RON might include more information, such as types, than CSV.

**Comment by XAMPPRocky** [comment]
This has been addressed in the latest version.
